### DISCUSSION TITLE
Make a example for tv-b-gone

### DISCUSSION THREAD

**Discussion by DjordjeMandic** [Arduino-IRremote/Arduino-IRremote#802] [Ideas]
Can anyone port tv-b-gone's world_ir_codes to be used with this library. For example just the NEC code on tvbgone takes 8bytes compresses + timings while the actual NEC code is 4 bytes. Could someone make a example sketch using library's functions like sendnec and etc... to transmit all codes from world_ir_codes? I'm working on a diy project board that's basically a mid power ir transmitter. Is it possible to create arrays of mixed codes to be sent using this library. I'm sure there will be flash space savings. It's also easier to read codes using PulsevlView and just write the address and command bytes of nec rather than writing full 8 bytes required by original tv-b-gone compression method.

**Comment by DjordjeMandic** [comment]
Alright i see, i should started discussion instead of issue. Alright since we have data pairs and timing pairs in flash already in tvbgone firmware can we somehow decompress every single code into ram and then send it as raw data using sendraw command.. I see that i got the answer to my qustion by myself. I will try it and if i succeed i will share it here

**Reply by ArminJo** [reply]
This would be very nice. Thanks!

**Comment by DjordjeMandic** [comment]
Alright here is what i did. 

Lets take a look at compressed IrCode struct:
```
struct IrCode {
  uint8_t timer_val;
  uint8_t numpairs;
  uint8_t bitcompression;
  uint16_t const *times;
  uint8_t const *codes;
};
```
The `numpairs` is 8bit number with maximum value of 256. That means we can have 256 different pairs of 2 `uint16_t` 16bit numbers;
Based on that we should create and allocate 1kB(256max pairs, each pair 4 bytes - 2bytes on time, 2bytes off time) of RAM for timing pairs. I did it this way so there is no way to run out of ram by having stack and heap collide.

So allocating the memory is done by creating 2 dimensional array of uint16_t;
`uint16_t TimingPairsBuffer[256][2]`

In loop where decompression takes place we can put timing pairs into buffer instead of emitting each pair and then emit that pair as raw data directly from ram. (Code taken from [arduino tv-b-gone 20161022 v1.3](https://github.com/shirriff/Arduino-TV-B-Gone))
```
for (uint8_t k=0; k<numpairs; k++) {
      uint16_t ti;
      //decompression takes around 25uS on average per pair @ 8mhz
      
      // Read the next 'n' bits as indicated by the compression variable
      // The multiply by 4 because there are 2 timing numbers per pair
      // and each timing number is one word long, so 4 bytes total!
      ti = (Read_bits(bitcompression)) * 4;

      // read the onTime and offTime from the program memory
      ontime = pgm_read_word(time_ptr+ti);  // read word 1 - ontime
      offtime = pgm_read_word(time_ptr+ti+2);  // read word 2 - offtime
      
      TimingPairsBuffer[k][0] = ontime;
      TimingPairsBuffer[k][1] = offtime;

      //we should transmit codes directly from ram in another for loop with interrupts disabled so that
      //timing is right and thus we get more precise off time, if we emmitted element right here
      //we would get higher off time than expected due to time required to decompress timing pair
      //XmitCodeElement(ontime, offtime, (freq!=0)); 
    }
    //Flush remaining bits, so that next code starts
    //with a fresh set of 8 bits.
    bitsleft_r=0;
```

This code is for emitting the code in arduino tv-b-gone firmware.

```
    bool pwmCode = (freq != 0);
    cli();  //disable interrupts so our custom delay in XmitCodeElement does not drift
    
    //emminting ir code from ram
    for(int tp = 0; tp < numpairs; tp++)
    {
      XmitCodeElement(TimingPairsBuffer[tp][0], TimingPairsBuffer[tp][1], pwmCode);
    }

    sei(); //enable interrupts
```
There we have improved off time accuraccy. 

That above was one approach using arduino tv-b-gone. We now have all timing pairs in ram but they are expressed in tens of uS. For example if were at `TimingPairsBuffer[0][0]` and lets imagine it returns 26. 26 actually means 26 tens of uS = 260uS.

All left to do is to write function for this library that will take `IrCode` struct and the TimingPairsBuffer by reference so it can decompress the code into user defined buffer first then transmit all. 

Function that takes only `IrCode` struct can also be added and the function will send timing pairs while decompressing at same time if off time accuracy is not a concern.

What do you think of this idea?

**Comment by DjordjeMandic** [comment]
Here are few screenshots of timings captured using Cypress FX2 based 8 channel 24mhz logic analyzer with sampling rate of 16MHz in sigrok PulseView that ATMega328P spitted out at 8MHz clock.

![image](https://user-images.githubusercontent.com/6750655/108874458-74664580-75fc-11eb-98f9-a7fa0dad7274.png)

![image](https://user-images.githubusercontent.com/6750655/108874662-a7103e00-75fc-11eb-832b-89abbae37b4e.png)

![image](https://user-images.githubusercontent.com/6750655/108874707-b3949680-75fc-11eb-8a2b-668a9f5d79d6.png)

![image](https://user-images.githubusercontent.com/6750655/108874739-bf805880-75fc-11eb-98c1-28e6fedf5fa1.png)


Here is few captures of NEC code that ATMega328P spitted out at 8MHz clock. I set pwmCode to false just so i can measure timing a bit easier. The off time is still off a little bit but much less compared to off time if code was decompressed at same time it was pulsed out. Here we can see that the for loop for sending data directly from ram actually increases off time by around 3uS. My delay function actually delays 562uS instead of 560uS but thats only 0.38% error. In off time we can see that its actually 566uS so the for loop for emitting code directly from ram only adds around 3uS at 8MHz clock. But if we were decompressing while emitting it would at at least 30uS to off time which is around 5.1% error at clock speed of 8MHz. Also the `PWM_code` for `XmitCodeElement` should be calculated before loop for emitting code since its constant for all pairs thus were not waisting cpu cycles and increasing off time just for comparing frequency with 0 for every element. It can be done before actually transmitting it and saved in ram as a single boolean and just passed to the funciton.

The longer the pulse is the more precise it is. For 560uS we get 562uS so its around 0.36% error and for 1670uS we get 1676uS so its around 0.36% error(in my timing pair definition this was set to 167 instead of 169). For 4500uS we get 4506uS so its around 0.13% error and for 9000uS we get 9003uS so its around 0.034% error

**Reply by DjordjeMandic** [reply]
Here is fixed 1690uS pulse.

![image](https://user-images.githubusercontent.com/6750655/108876448-82b56100-75fe-11eb-9820-5b892c98128c.png)

![image](https://user-images.githubusercontent.com/6750655/108876555-a082c600-75fe-11eb-863a-5821cdbee60e.png)

`NEC_LOGIC_ONE_LEN` was set to 167 instead of 169
